sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/ScalaReflectionSuite.scala

```diff
@@ -109,6 +109,64 @@ object TestingUDT {
   }
 }
 
+/** An example derived from Twitter/Scrooge codegen for thrift  */
+object ScroogeLikeExample {
+  def apply(
+    x: Int
+  ): ScroogeLikeExample =
+    new Immutable(
+      x
+    )
+
+  def unapply(_item: ScroogeLikeExample): _root_.scala.Option[Int] = _root_.scala.Some(_item.x)
+
+  class Immutable(val x: Int) extends ScroogeLikeExample
+
+  trait Proxy extends ScroogeLikeExample {
+    protected def _underlying_ScroogeLikeExample: ScroogeLikeExample
+    override def x: Int = _underlying_ScroogeLikeExample.x
+  }
+}
+
+trait ScroogeLikeExample
+  extends _root_.scala.Product1[Int]
```

Comment:
Does it really need to extend Product1 for the test? quite possibly, just didn't see it.

Reply:
It provides `productArity` used in `equals` used below.